mitxgraders/formulagrader/integralgrader.py

```diff
@@ -537,3 +551,326 @@ def raw_integrand(x):
+                varscope=varlist,
+                funcscope=funclist
+            )
+
+            # Save results
+            instructor_evals.append(expected_eval)
+            student_evals.append(student_eval)
+
+            # Put the instructor variables back in for the debug output
+            varlist.update(var_samples[i])
+            self.log_eval_info(i, varlist, funclist,
+                               student_eval=student_eval,
+                               instructor_eval=expected_eval)
+
+        return instructor_evals, student_evals, used_funcs
+
+    def evaluate_sum(self, summand_str, lower_str, upper_str, summation_var,
+                     varscope=None, funcscope=None):
+        varscope = {} if varscope is None else varscope
+        funcscope = {} if funcscope is None else funcscope
+
+        # Unlike integration, we do not allow the summation variable to appear in
+        # the limits of the sum. Make sure the summation variable is not defined
+        # in the varscope. Note that this means that i and j cannot be used as
+        # summation variables.
+        if summation_var in varscope:
+            msg = 'Summation variable {} conflicts with another previously-defined variable.'
+            raise SummationError(msg.format(summation_var))
+
+        # Evaluate the limits, and find the functions that will be used.
+        lower, upper, used_funcs = self.get_limits_and_funcs(summand_str, lower_str, upper_str,
+                                                             summation_var, varscope, funcscope)
+
+        # Check to ensure that sum limits are not complex.
+        if isinstance(lower, complex) or isinstance(upper, complex):
+            raise SummationError('Summation limits must be real but have evaluated '
+                                 'to complex numbers.')
+
+        # Check to ensure that sum limits are integers or infinite
+        if abs(lower) != float('inf') and int(lower) != lower:
+            raise SummationError('Lower summation limit does not evaluate to an integer.')
+        if abs(upper) != float('inf') and int(upper) != upper:
+            raise SummationError('Upper summation limit does not evaluate to an integer.')
+
+        def eval_summand(x):
+            """
+            Helper function to evaluate the summand at the given value of the
+            summation variable.
+            """
+            varscope[summation_var] = x
+            value, _ = evaluator(summand_str,
+                                 variables=varscope,
+                                 functions=funcscope,
+                                 suffixes=self.suffixes)
+            del varscope[summation_var]
+            return value
+
+        # Check if used_funcs includes a factorial function
+        if 'fact' in used_funcs or 'factorial' in used_funcs:
+            infty_val = self.config['infty_val_fact']
+        else:
+            infty_val = self.config['infty_val']
```

Comment:
**Question:** Is `infty_val_fact` really worthwhile? Authors could just set `infty_val` to be small themselves. My main "concern" would just be that this gives a false sense of security.... there are super-factorial terms that are at least somewhat common. E.g., sums in calculus textbooks often have terms like `n^n`.

But... I am not particularly familiar with the actual problems people will be writing with this. So maybe it's totally reasonable.

Reply:
The idea is to try to detect when an overflow/underflow is going to occur in an infinite summation, and use a smaller number to prevent it from happening. I'd prefer that students not get overflow messages that are confusing. The most common source of that will be from factorials, I suspect, which saves the problem author from having to worry about setting `infty_val` each time. I'm not sure that superfactorial type terms will show up all that often in an infinite summation, no? Whereas factorials are very, very common in infinite summations.

Reply:
you're probably right that this will show up much more often, and that makes sense.

Superfactorial terms like `n^n` show up a lot in calculus books when doing convergence tests (i.e., does it converge, not "what does it convert to"). But I dunno if they'll show up in SumGrader problems.